controllers/awsmachine_controller.go

```diff
@@ -384,28 +412,40 @@ func (r *AWSMachineReconciler) reconcileNormal(_ context.Context, machineScope *
 
 	if !machineScope.Cluster.Status.InfrastructureReady {
 		machineScope.Info("Cluster infrastructure is not ready yet")
+		conditions.MarkFalse(machineScope.AWSMachine, infrav1.InstanceReadyCondition, infrav1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
 		return ctrl.Result{}, nil
 	}
 
 	// Make sure bootstrap data is available and populated.
 	if machineScope.Machine.Spec.Bootstrap.DataSecretName == nil {
 		machineScope.Info("Bootstrap data secret reference is not yet available")
+		conditions.MarkFalse(machineScope.AWSMachine, infrav1.InstanceReadyCondition, infrav1.WaitingForBootstrapDataReason, clusterv1.ConditionSeverityInfo, "")
 		return ctrl.Result{}, nil
 	}
 
 	ec2svc := r.getEC2Service(clusterScope)
 
-	// Get or create the instance.
-	instance, err := r.getOrCreate(machineScope, ec2svc, secretSvc)
+	// Find existing instance
+	instance, err := r.findInstance(machineScope, ec2svc)
 	if err != nil {
+		conditions.MarkFalse(machineScope.AWSMachine, infrav1.InstanceReadyCondition, infrav1.InstanceNotFoundReason, clusterv1.ConditionSeverityError, err.Error())
 		return ctrl.Result{}, err
 	}
+	// Create new instance
+	if instance == nil {
```

Comment:
I think I found a potential pre-existing bug...  It looks like we are swallowing up errors when we attempt to do a lookup for a specific instance ID (`ec2svc.InstanceIfExists()`), when the AWS api returns a NotFound error. I'm not sure that is the behavior that we want, since:
- It introduces a potential re-entrancy bug related to the eventual consistency of the AWS APIs, where we could potentially create more than one instance for a given machine
- It could potentially lead to the replacement of an instance that was terminated and purged, when it shouldn't.

Reply:
Extracted this comment to https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1769